File: android/app/src/main/java/se/iloppis/app/ui/screens/cashier/CashierScreen.kt
package se.iloppis.app.ui.screens.cashier

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.*
import androidx.activity.compose.BackHandler
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import se.iloppis.app.R
import se.iloppis.app.domain.model.Event
import se.iloppis.app.ui.dialogs.InvalidSellerDialog
import se.iloppis.app.ui.dialogs.ServerErrorDialog
import se.iloppis.app.ui.screens.pending.PendingPurchasesScreen
import se.iloppis.app.ui.screens.review.PurchaseReviewScreen
import se.iloppis.app.ui.components.buttons.AppButton
import se.iloppis.app.ui.components.buttons.AppButtonSize
import se.iloppis.app.ui.components.buttons.AppButtonVariant
import se.iloppis.app.ui.theme.AppColors

/**
 * Main cashier screen with numeric keypad, transaction list, and payment options.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CashierScreen(
    event: Event,
    apiKey: String,
    onBack: () -> Unit
) {
    val viewModel: CashierViewModel = viewModel(
        key = "cashier-${event.id}",
        factory = CashierViewModel.factory(
            eventId = event.id,
            eventName = event.name,
            apiKey = apiKey
        )
    )
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()

    var showPendingInfoDialog by remember { mutableStateOf(false) }
    var showClosePendingDialog by remember { mutableStateOf(false) }
    var showReviewScreen by remember { mutableStateOf(false) }
    var showDetailedReview by remember { mutableStateOf<String?>(null) }
    var closeRequested by remember { mutableStateOf(false) }

    if (showClosePendingDialog) {
        AlertDialog(
            onDismissRequest = { showClosePendingDialog = false },
            confirmButton = {
                AppButton(
                    text = stringResource(R.string.cashier_close_pending_confirm),
                    onClick = {
                        showClosePendingDialog = false
                        closeRequested = true
                    },
                    variant = AppButtonVariant.Warning,
                    size = AppButtonSize.Small
                )
            },
            dismissButton = {
                AppButton(
                    text = stringResource(R.string.button_cancel),
                    onClick = { showClosePendingDialog = false },
                    variant = AppButtonVariant.Text,
                    size = AppButtonSize.Small
                )
            },
            title = { Text(stringResource(R.string.cashier_close_pending_title)) },
            text = {
                Text(
                    stringResource(
                        R.string.cashier_close_pending_message,
                        uiState.pendingSoldItemsCount,
                    )
                )
            }
        )
    }

    // Show detailed purchase review if requested
    if (showDetailedReview != null) {
        val detailedViewModel: se.iloppis.app.ui.screens.review.DetailedPurchaseReviewViewModel = viewModel(
            key = "detailed-$showDetailedReview",
            factory = se.iloppis.app.ui.screens.review.DetailedPurchaseReviewViewModel.factory(
                purchaseId = showDetailedReview!!,
                eventId = event.id,
                apiKey = apiKey
            )
        )
        se.iloppis.app.ui.screens.review.DetailedPurchaseReviewScreen(
            viewModel = detailedViewModel,
            onBack = {
                showDetailedReview = null
                // Refresh rejected purchases list
                viewModel.refreshRejectedPurchasesCount()
            }
        )
        return
    }

    // Show pending purchases screen if requested
    if (showReviewScreen) {
        PendingPurchasesScreen(
            apiKey = apiKey,
            eventId = event.id,
            onNavigateBack = { showReviewScreen = false }
        )
        return
    }

    LaunchedEffect(closeRequested) {
        if (!closeRequested) return@LaunchedEffect
        val closeSucceeded = viewModel.requestCloseAndFlush()
        if (closeSucceeded) {
            onBack()
        } else {
            closeRequested = false
        }
    }

    BackHandler {
        if (closeRequested) {
            return@BackHandler
        }
        if (uiState.pendingSoldItemsCount > 0) {
            showClosePendingDialog = true
        } else {
            closeRequested = true
        }
    }

    if (showPendingInfoDialog) {
        AlertDialog(
            onDismissRequest = { showPendingInfoDialog = false },
            confirmButton = {
                AppButton(
                    text = stringResource(R.string.common_ok),
                    onClick = { showPendingInfoDialog = false },
                    variant = AppButtonVariant.Text,
                    size = AppButtonSize.Small
                )
            },
            title = { Text(stringResource(R.string.cashier_pending_info_title)) },
            text = {
                Text(
                    stringResource(R.string.cashier_pending_info_message)
                )
            }
        )
    }

    // Server error dialog
    if (uiState.showServerErrorDialog) {
        ServerErrorDialog(
            pendingPurchasesCount = uiState.rejectedPurchasesCount,
            onDismiss = { viewModel.onAction(CashierAction.DismissServerErrorDialog) }
        )
    }

    // Invalid seller dialog
    val dialogData = uiState.invalidSellerDialogData
    if (uiState.showInvalidSellerDialog && dialogData != null) {
        InvalidSellerDialog(
            purchaseId = dialogData.purchaseId,
            timestamp = dialogData.timestamp,
            invalidSellers = dialogData.invalidSellers,
            onDismiss = { viewModel.onAction(CashierAction.DismissInvalidSellerDialog) },
            onReviewNow = {
                viewModel.onAction(CashierAction.DismissInvalidSellerDialog)
                showReviewScreen = true
            }
        )
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = {
                    Column {
                        Text(stringResource(R.string.cashier_title))
                        Text(
                            text = event.name,
                            fontSize = 12.sp,
                            color = AppColors.DialogBackground.copy(alpha = 0.8f),
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis
                        )
                    }
                },
                actions = {
                    // Show rejected purchases badge
                    if (uiState.rejectedPurchasesCount > 0) {
                        AppButton(
                            text = uiState.rejectedPurchasesCount.toString(),
                            onClick = { showReviewScreen = true },
                            variant = AppButtonVariant.Text,
                            contentColor = AppColors.DialogBackground,
                            size = AppButtonSize.Small,
                            leadingIcon = {
                                Icon(
                                    imageVector = Icons.Default.Warning,
                                    contentDescription = stringResource(R.string.pending_purchases_title)
                                )
                            }
                        )
                        Spacer(modifier = Modifier.width(8.dp))
                    }

                    if (uiState.pendingSoldItemsCount > 0) {
                        IconButton(onClick = { showPendingInfoDialog = true }) {
                            Icon(
                                imageVector = Icons.Filled.Info,
                                contentDescription = stringResource(R.string.content_description_pending_info),
                                tint = AppColors.DialogBackground
                            )
                        }
                        Spacer(modifier = Modifier.width(8.dp))
                        Text(
                            text = uiState.pendingSoldItemsCount.toString(),
                            color = AppColors.DialogBackground,
                            fontWeight = FontWeight.Bold
                        )
                        Spacer(modifier = Modifier.width(12.dp))
                    }
                },
                navigationIcon = {
                    IconButton(onClick = {
                        if (closeRequested) {
                            return@IconButton
                        }
                        if (uiState.pendingSoldItemsCount > 0) {
                            showClosePendingDialog = true
                        } else {
                            closeRequested = true
                        }
                    }) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                            contentDescription = stringResource(R.string.button_back)
                        )
                    }
                },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = AppColors.ButtonSuccess,
                    titleContentColor = AppColors.DialogBackground,
                    navigationIconContentColor = AppColors.DialogBackground
                )
            )
        },
        snackbarHost = {
            // Show warnings and errors
            uiState.warningMessage?.let { message ->
                Snackbar(
                    modifier = Modifier.padding(16.dp),
                    action = {
                        AppButton(
                            text = stringResource(R.string.common_ok),
                            onClick = { viewModel.onAction(CashierAction.DismissWarning) },
                            variant = AppButtonVariant.Text,
                            contentColor = AppColors.OnButtonPrimary,
                            size = AppButtonSize.Small
                        )
                    },
                    containerColor = AppColors.Warning
                ) {
                    Text(message.asString())
                }
            }
            uiState.errorMessage?.let { message ->
                Snackbar(
                    modifier = Modifier.padding(16.dp),
                    action = {
                        AppButton(
                            text = stringResource(R.string.common_ok),
                            onClick = { viewModel.onAction(CashierAction.DismissError) },
                            variant = AppButtonVariant.Text,
                            contentColor = AppColors.OnButtonPrimary,
                            size = AppButtonSize.Small
                        )
                    },
                    containerColor = AppColors.Error
                ) {
                    Text(message.asString())
                }
            }
        }
    ) { padding ->
        if (uiState.isLoading) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(padding),
                contentAlignment = Alignment.Center
            ) {
                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(16.dp)
                ) {
                    CircularProgressIndicator(color = AppColors.Primary)
                    Text(
                        text = stringResource(R.string.cashier_loading),
                        color = AppColors.TextMuted
                    )
                }
            }
        } else {
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(padding)
                    .padding(horizontal = 16.dp)
                    .verticalScroll(rememberScrollState()),
                verticalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                Spacer(modifier = Modifier.height(8.dp))

                // Input fields for seller number and price
                InputSection(
                    sellerNumber = uiState.sellerNumber,
                    priceString = uiState.priceString,
                    activeField = uiState.activeField,
                    onSellerClick = { viewModel.onAction(CashierAction.SetActiveField(ActiveField.SELLER)) },
                    onPriceClick = { viewModel.onAction(CashierAction.SetActiveField(ActiveField.PRICE)) }
                )

                // Numeric keypad
                NumericKeypad(
                    onDigitPress = { viewModel.onAction(CashierAction.KeypadPress(it)) },
                    onClear = { viewModel.onAction(CashierAction.KeypadClear) },
                    onBackspace = { viewModel.onAction(CashierAction.KeypadBackspace) },
                    onSpace = { viewModel.onAction(CashierAction.KeypadSpace) },
                    onOk = { viewModel.onAction(CashierAction.KeypadOk) }
                )

                // Transaction list
                TransactionList(
                    transactions = uiState.transactions,
                    onRemoveItem = { viewModel.onAction(CashierAction.RemoveItem(it)) },
                    onClearAll = { viewModel.onAction(CashierAction.ClearAllItems) }
                )

                // Payment section
                PaymentSection(
                    total = uiState.total,
                    paidAmount = uiState.paidAmount,
                    change = uiState.change,
                    isProcessing = uiState.isProcessingPayment,
                    onPaidAmountChange = { viewModel.onAction(CashierAction.SetPaidAmount(it)) },
                    onCashPayment = { viewModel.onAction(CashierAction.Checkout(PaymentMethodType.CASH)) },
                    onSwishPayment = { viewModel.onAction(CashierAction.Checkout(PaymentMethodType.SWISH)) }
                )

                Spacer(modifier = Modifier.height(16.dp))
            }
        }
    }
}

@Composable
private fun InputSection(
    sellerNumber: String,
    priceString: String,
    activeField: ActiveField,
    onSellerClick: () -> Unit,
    onPriceClick: () -> Unit
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        // Seller number input
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = stringResource(R.string.cashier_seller_label),
                fontSize = 12.sp,
                color = AppColors.TextMuted,
                modifier = Modifier.padding(bottom = 4.dp)
            )
            InputField(
                value = sellerNumber,
                placeholder = "123",
                isActive = activeField == ActiveField.SELLER,
                onClick = onSellerClick
            )
        }

        // Price input
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = stringResource(R.string.cashier_price_label),
                fontSize = 12.sp,
                color = AppColors.TextMuted,
                modifier = Modifier.padding(bottom = 4.dp)
            )
            InputField(
                value = priceString,
                placeholder = "50 100",
                isActive = activeField == ActiveField.PRICE,
                onClick = onPriceClick
            )
        }
    }
}

@Composable
private fun InputField(
    value: String,
    placeholder: String,
    isActive: Boolean,
    onClick: () -> Unit
) {
    val borderColor = if (isActive) AppColors.Primary else AppColors.Border
    val backgroundColor = if (isActive) AppColors.Primary.copy(alpha = 0.05f) else AppColors.CardBackground

    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(48.dp)
            .clip(RoundedCornerShape(8.dp))
            .background(backgroundColor)
            .border(
                width = if (isActive) 2.dp else 1.dp,
                color = borderColor,
                shape = RoundedCornerShape(8.dp)
            )
            .clickable(onClick = onClick)
            .padding(horizontal = 12.dp),
        contentAlignment = Alignment.CenterStart
    ) {
        if (value.isEmpty()) {
            Text(
                text = placeholder,
                color = AppColors.TextMuted.copy(alpha = 0.5f),
                fontSize = 16.sp
            )
        } else {
            Text(
                text = value,
                color = AppColors.TextPrimary,
                fontSize = 16.sp,
                fontWeight = FontWeight.Medium
            )
        }
    }
}

File: ios/iLoppis/UI/Screens/Cashier/CashierScreen.swift
import SwiftUI

struct CashierScreen: View {
    let event: Event
    let apiKey: String
    let onBack: () -> Void

    @StateObject private var viewModel: CashierViewModel

    init(event: Event, apiKey: String, onBack: @escaping () -> Void) {
        self.event = event
        self.apiKey = apiKey
        self.onBack = onBack
        _viewModel = StateObject(
            wrappedValue: CashierViewModel(
                eventId: event.id,
                eventName: event.name,
                apiKey: apiKey
            )
        )
    }

    var body: some View {
        VStack(spacing: 0) {
            header
            ScrollView {
                VStack(alignment: .leading, spacing: 12) {
                    if viewModel.state.isLoading {
                        Text(LocalizedStringKey("cashier_loading"))
                            .foregroundColor(AppColors.textSecondary)
                    }

                    inputSection
                    itemsSection
                    totalsSection
                    paymentSection
                }
                .padding(16)
            }
        }
        .background(AppColors.background.ignoresSafeArea())
        .alert(
            LocalizedStringKey("app_title"),
            isPresented: Binding(
                get: { viewModel.state.warningMessage != nil },
                set: { if !$0 { viewModel.onAction(.dismissWarning) } }
            ),
            actions: {
                Button(LocalizedStringKey("button_cancel")) {
                    viewModel.onAction(.dismissWarning)
                }
            },
            message: {
                Text(viewModel.state.warningMessage ?? "")
            }
        )
        .alert(
            LocalizedStringKey("app_title"),
            isPresented: Binding(
                get: { viewModel.state.errorMessage != nil },
                set: { if !$0 { viewModel.onAction(.dismissError) } }
            ),
            actions: {
                Button(LocalizedStringKey("button_cancel")) {
                    viewModel.onAction(.dismissError)
                }
            },
            message: {
                Text(viewModel.state.errorMessage ?? "")
            }
        )
        .overlay {
            if viewModel.state.isProcessingPayment {
                ZStack {
                    Color.black.opacity(0.3).ignoresSafeArea()
                    ProgressView()
                        .tint(.white)
                        .scaleEffect(1.2)
                }
            }
        }
    }

    private var header: some View {
        HStack {
            Button(action: onBack) {
                Image(systemName: "chevron.backward")
                    .foregroundColor(AppColors.textPrimary)
            }
            Text(LocalizedStringKey("cashier_title"))
                .font(.title2.weight(.bold))
                .foregroundColor(AppColors.textPrimary)
            Spacer()
        }
        .padding(16)
        .background(AppColors.background)
    }

    private var inputSection: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text(String(format: NSLocalizedString("cashier_event_label", comment: ""), event.name))
                .font(.footnote)
                .foregroundColor(AppColors.textMuted)

            HStack(spacing: 12) {
                fieldButton(
                    titleKey: "cashier_seller_label",
                    value: viewModel.state.sellerNumber,
                    isActive: viewModel.state.activeField == .seller
                ) {
                    viewModel.onAction(.setActiveField(.seller))
                }
                fieldButton(
                    titleKey: "cashier_price_label",
                    value: viewModel.state.priceString,
                    isActive: viewModel.state.activeField == .price
                ) {
                    viewModel.onAction(.setActiveField(.price))
                }
            }

            keypad
        }
        .padding(16)
        .background(AppColors.cardBackground)
        .cornerRadius(12)
        .overlay(
            RoundedRectangle(cornerRadius: 12)
                .stroke(AppColors.border, lineWidth: 1)
        )
    }

    private func fieldButton(titleKey: String, value: String, isActive: Bool, onTap: @escaping () -> Void) -> some View {
        Button(action: onTap) {
            VStack(alignment: .leading, spacing: 4) {
                Text(LocalizedStringKey(titleKey))
                    .font(.caption)
                    .foregroundColor(AppColors.textMuted)
                Text(value.isEmpty ? "—" : value)
                    .font(.headline)
                    .foregroundColor(AppColors.textPrimary)
                    .lineLimit(1)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(12)
            .background(isActive ? AppColors.badgeUpcomingBackground : AppColors.background)
            .cornerRadius(10)
            .overlay(
                RoundedRectangle(cornerRadius: 10)
                    .stroke(isActive ? AppColors.buttonPrimary : AppColors.border, lineWidth: 1)
            )
        }
        .buttonStyle(.plain)
    }

    private var keypad: some View {
        VStack(spacing: 10) {
            ForEach([["1","2","3"],["4","5","6"],["7","8","9"],["C","0","OK"]], id: \.self) { row in
                HStack(spacing: 10) {
                    ForEach(row, id: \.self) { key in
                        keypadButton(title: key)
                    }
                }
            }
            HStack(spacing: 10) {
                keypadButton(title: "⌫")
                keypadButton(title: "␣")
                keypadButton(title: "+")
            }
        }
    }

    private func keypadButton(title: String) -> some View {
        Button {
            switch title {
            case "C":
                viewModel.onAction(.keypadClear)
            case "OK":
                viewModel.onAction(.keypadOk)
            case "⌫":
                viewModel.onAction(.keypadBackspace)
            case "␣":
                viewModel.onAction(.keypadSpace)
            case "+":
                viewModel.onAction(.keypadPress(" "))
            default:
                viewModel.onAction(.keypadPress(title))
            }
        } label: {
            Text(title)
                .font(.headline)
                .frame(maxWidth: .infinity)
                .padding(.vertical, 14)
                .foregroundColor(AppColors.textPrimary)
                .background(AppColors.background)
                .cornerRadius(10)
                .overlay(
                    RoundedRectangle(cornerRadius: 10)
                        .stroke(AppColors.border, lineWidth: 1)
                )
        }
        .buttonStyle(.plain)
    }

    private var itemsSection: some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack {
                Text(LocalizedStringKey("cashier_items_title"))
                    .font(.headline)
                    .foregroundColor(AppColors.textPrimary)
                Spacer()
                Button(action: { viewModel.onAction(.clearAllItems) }) {
                    Text(LocalizedStringKey("cashier_clear_all"))
                        .foregroundColor(AppColors.textSecondary)
                }
                .buttonStyle(.plain)
                .disabled(viewModel.state.transactions.isEmpty)
            }

            if viewModel.state.transactions.isEmpty {
                Text(LocalizedStringKey("cashier_no_items"))
                    .foregroundColor(AppColors.textSecondary)
            } else {
                VStack(spacing: 8) {
                    ForEach(viewModel.state.transactions) { item in
                        HStack {
                            Text("\(NSLocalizedString("cashier_seller_header", comment: "")) \(item.sellerNumber)")
                                .foregroundColor(AppColors.textPrimary)
                            Spacer()
                            Text("\(item.price)")
                                .foregroundColor(AppColors.textPrimary)
                            Button(action: { viewModel.onAction(.removeItem(item.id)) }) {
                                Text(LocalizedStringKey("cashier_remove_item"))
                                    .font(.caption)
                                    .foregroundColor(AppColors.textMuted)
                                    .padding(.leading, 8)
                            }
                            .buttonStyle(.plain)
                        }
                        .padding(12)
                        .background(AppColors.cardBackground)
                        .cornerRadius(10)
                        .overlay(
                            RoundedRectangle(cornerRadius: 10)
                                .stroke(AppColors.border, lineWidth: 1)
                        )
                    }
                }
            }
        }
        .padding(16)
        .background(AppColors.cardBackground)
        .cornerRadius(12)
        .overlay(
            RoundedRectangle(cornerRadius: 12)
                .stroke(AppColors.border, lineWidth: 1)
        )
    }

    private var totalsSection: some View {
        VStack(spacing: 10) {
            row(labelKey: "cashier_total", value: "\(viewModel.state.total)")
            HStack {
                Text(LocalizedStringKey("cashier_paid"))
                    .foregroundColor(AppColors.textSecondary)
                Spacer()
                TextField("0", text: Binding(
                    get: { viewModel.state.paidAmount },
                    set: { viewModel.onAction(.setPaidAmount($0)) }
                ))
                .keyboardType(.numberPad)
                .multilineTextAlignment(.trailing)
                .frame(width: 100)
                .padding(8)
                .background(AppColors.background)
                .cornerRadius(8)
            }
            row(labelKey: "cashier_change", value: "\(viewModel.state.change)")
        }
        .padding(16)
        .background(AppColors.cardBackground)
        .cornerRadius(12)
        .overlay(
            RoundedRectangle(cornerRadius: 12)
                .stroke(AppColors.border, lineWidth: 1)
        )
    }

    private func row(labelKey: String, value: String) -> some View {
        HStack {
            Text(LocalizedStringKey(labelKey))
                .foregroundColor(AppColors.textSecondary)
            Spacer()
            Text(value)
                .foregroundColor(AppColors.textPrimary)
        }
    }

    private var paymentSection: some View {
        HStack(spacing: 12) {
            Button {
                viewModel.onAction(.checkout(.cash))
            } label: {
                Text(LocalizedStringKey("cashier_button_cash"))
                    .frame(maxWidth: .infinity)
                    .padding()
                    .foregroundColor(.white)
                    .background(AppColors.buttonPrimary)
                    .cornerRadius(10)
            }
            .buttonStyle(.plain)
            .disabled(viewModel.state.transactions.isEmpty)

            Button {
                viewModel.onAction(.checkout(.swish))
            } label: {
                Text(LocalizedStringKey("cashier_button_swish"))
                    .frame(maxWidth: .infinity)
                    .padding()
                    .foregroundColor(.white)
                    .background(AppColors.buttonPrimary)
                    .cornerRadius(10)
            }
            .buttonStyle(.plain)
            .disabled(viewModel.state.transactions.isEmpty)
        }
        .padding(.bottom, 8)
    }
}
import SwiftUI
